### PR TITLE
Fix sklearn pipeline predict proba function 

### LINKED ISSUE #12445: [BUG] Logged sklearn models pyfunc flavour does not work when using sklearn pipelines on sklearn>=1.4+
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

- Client: 2.14.1

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.03
- **Python version**: 3.12.3

### Describe the problem

`mlflow.models.predict` on a logged sklearn model pipeline fails if you use `pyfunc_predict_fn="predict_proba"` in combination with a sklearn pipeline:

```
TypeError: HistGradientBoostingClassifier.predict_proba() got an unexpected keyword argument 'params'
```

### Tracking information

-

### Code to reproduce issue

```python
import pandas as pd
from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.datasets import make_hastie_10_2
from sklearn.pipeline import make_pipeline

# train the model
X, y = make_hastie_10_2(random_state=0)
X = pd.DataFrame(X, columns=[f"f{i}" for i in range(X.shape[1])])
X_train, X_test = X[:2000], X[2000:]
y_train, y_test = y[:2000], y[2000:]
# notice we use a sklearn pipeline. Without it would work fine!
clf = make_pipeline(HistGradientBoostingClassifier(max_iter=100)).fit(X_train, y_train)

# log the model
with mlflow.start_run():
    mlflow.sklearn.log_model(
            sk_model=clf,
            artifact_path="model",
            code_paths=None,
            signature=None,
            input_example=X_train.sample(1),
            pyfunc_predict_fn="predict_proba", # <-- Notice it would work fine with the default "predict"
            pip_requirements=["scikit-learn==1.5.0"],
            registered_model_name="bug_report",
    )

# Assert predictions work
assert clf.predict_proba(X_test.sample(1)) is not None

# Run the model
mlflow.models.predict(
    model_uri="models:/bug_report/1",
    input_data=X_test.sample(1),
    env_manager="local",
)
#> TypeError: HistGradientBoostingClassifier.predict_proba() got an unexpected keyword argument 'params'
```

### Stack trace

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 mlflow.models.predict(
      2     model_uri="models:/test_tim/15",
      3     input_data=X_test.sample(1),
      4     env_manager="local",
      5 )

File /anaconda/envs/ds_ramen/lib/python3.11/site-packages/mlflow/models/python_api.py:180, in predict(model_uri, input_data, input_path, content_type, output_path, env_manager, install_mlflow, pip_requirements_override)
    177         with open(input_path, "w") as f:
    178             f.write(input_data)
--> 180         _predict(input_path)
    181 else:
    182     _predict(input_path)

File /anaconda/envs/ds_ramen/lib/python3.11/site-packages/mlflow/models/python_api.py:159, in predict.<locals>._predict(_input_path)
    156 def _predict(_input_path: str):
    157     return get_flavor_backend(
    158         model_uri, env_manager=env_manager, install_mlflow=install_mlflow
--> 159     ).predict(
    160         model_uri=model_uri,
    161         input_path=_input_path,
    162         output_path=output_path,
    163         content_type=content_type,
    164         pip_requirements_override=pip_requirements_override,
    165     )

File /anaconda/envs/ds_ramen/lib/python3.11/site-packages/mlflow/pyfunc/backend.py:217, in PyFuncBackend.predict(self, model_uri, input_path, output_path, content_type, pip_requirements_override)
    212 if pip_requirements_override:
    213     raise MlflowException(
    214         "`pip_requirements_override` is not supported for local env manager."
    215         "Please use conda or virtualenv instead."
    216     )
--> 217 scoring_server._predict(local_uri, input_path, output_path, content_type)

File /anaconda/envs/ds_ramen/lib/python3.11/site-packages/mlflow/pyfunc/scoring_server/__init__.py:470, in _predict(model_uri, input_path, output_path, content_type)
    467     raise Exception(f"Unknown content type '{content_type}'")
    469 if inspect.signature(pyfunc_model.predict).parameters.get("params"):
--> 470     raw_predictions = pyfunc_model.predict(df, params=params)
    471 else:
    472     _log_warning_if_params_not_in_predict_signature(_logger, params)

File /anaconda/envs/ds_ramen/lib/python3.11/site-packages/mlflow/pyfunc/__init__.py:737, in PyFuncModel.predict(self, data, params)
    735 with self._try_get_or_generate_prediction_context() as context:
    736     self._update_dependencies_schemas_in_prediction_context(context)
--> 737     return self._predict(data, params)

File /anaconda/envs/ds_ramen/lib/python3.11/site-packages/mlflow/pyfunc/__init__.py:769, in PyFuncModel._predict(self, data, params)
    767 data, params = self._validate_prediction_input(data, params)
    768 if inspect.signature(self._predict_fn).parameters.get("params"):
--> 769     return self._predict_fn(data, params=params)
    771 _log_warning_if_params_not_in_predict_signature(_logger, params)
    772 return self._predict_fn(data)

File /anaconda/envs/ds_ramen/lib/python3.11/site-packages/sklearn/pipeline.py:724, in Pipeline.predict_proba(self, X, **params)
    722     for _, name, transform in self._iter(with_final=False):
    723         Xt = transform.transform(Xt)
--> 724     return self.steps[-1][1].predict_proba(Xt, **params)
    726 # metadata routing enabled
    727 routed_params = process_routing(self, "predict_proba", **params)

TypeError: HistGradientBoostingClassifier.predict_proba() got an unexpected keyword argument 'params'
```

### Other info / logs

This bit from the stacktrace is relevant:

```
    469 if inspect.signature(pyfunc_model.predict).parameters.get("params"):
--> 470     raw_predictions = pyfunc_model.predict(df, params=params)
```

Here's what going on:

```python
import inspect
from sklearn.ensemble import HistGradientBoostingClassifier
m = HistGradientBoostingClassifier()
assert inspect.signature(m.predict_proba).parameters.get("params") == None  # all OK!

from sklearn.pipeline import make_pipeline
m = make_pipeline(HistGradientBoostingClassifier())
assert inspect.signature(m.predict_proba).parameters.get("params") == None  # Fails! 
```

If you look at the docs for `Pipeline` ([here](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline.predict_proba)) you can see that `sklearn.pipeline.Pipeline.predict_proba` has a `**params` argument added since sklearn 1.4. 

```python
inspect.signature(m.predict_proba).parameters
#> mappingproxy({'X': <Parameter "X">, 'params': <Parameter "**params">})
```

I'm not sure how to fix, but it definitely seems like a bug to me. 

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by WeichenXu123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/WeichenXu123/mlflow/pull/12489?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12489/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12489
```

</p>
</details>

### Related Issues/PRs

 Closes #12445

### What changes are proposed in this pull request?

Fix sklearn pipeline predict proba function:

sklearn pipeline predict / predict_proba functions has `**params*` argument, the argument name accidentally matches mlflow pyfunc model predict method `params` argument, but they are different. 
I added logic in MLflow to handle this.

Note that this approach will cause `_model_impl` APIs breaking change,

the alternative approach https://github.com/mlflow/mlflow/pull/12554 doesn't have this issue.

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [X] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by BenWilson2** [comment]
Can we add an additional params-included test in the sklearn flavor suite to validate that predict_proba is capable of handling params like predict() is?

**Comment by WeichenXu123** [comment]
> Can we add an additional params-included test in the sklearn flavor suite to validate that predict_proba is capable of handling params like predict() is?

Added !

File: tests/sklearn/test_sklearn_model_export.py [hunk 1]
```diff
@@ -874,3 +876,47 @@ def test_model_registration_metadata_handling(sklearn_knn_model, tmp_path):
...
+    knn_model = sklearn_knn_model.model
+
+    predict_invocation_params = []
+    original_predict = knn_model.predict
+
+    def predict(data, p1=None):
+        predict_invocation_params.append(p1)
+        return original_predict(data)
+
+    knn_model.predict = predict
```
Comment (harupy) [inline tests/sklearn/test_sklearn_model_export.py:897]: why do we need do this?

Comment (harupy) [reply tests/sklearn/test_sklearn_model_export.py:897]: can we use the model in #12445?

**Comment by harupy** [comment]
I'm curious how `params` are passed in the pipeline `predict_proba` method.

File: mlflow/sklearn/__init__.py [hunk 2]
```diff
@@ -514,9 +524,10 @@ def __init__(self, sklearn_model):
 
         # Patch the model with custom predict functions that can be specified
         # via `pyfunc_predict_fn` argument when saving or logging.
-        for predict_fn in self._SUPPORTED_CUSTOM_PREDICT_FN:
-            if fn := getattr(self.sklearn_model, predict_fn, None):
-                setattr(self, predict_fn, fn)
+        for predict_fn_name in self._SUPPORTED_CUSTOM_PREDICT_FN:
+            if fn := getattr(self.sklearn_model, predict_fn_name, None):
+                predict_fn = _gen_sklearn_model_predict_fn(fn, predict_fn_name)
+                setattr(self, predict_fn_name, predict_fn)
```
Comment (harupy) [inline mlflow/sklearn/__init__.py:530]: we just might want to define each method for readbility.

```
def predict_proba(...):
    ...

def predict_log_proba(...):
    ...
```

File: tests/evaluate/test_default_evaluator.py [hunk 3]
```diff
@@ -187,7 +187,8 @@ def test_regressor_evaluation(
 
     expected_metrics = _get_regressor_metrics(y, y_pred, sample_weights=sample_weights)
     expected_metrics["score"] = model._model_impl.score(
-        diabetes_dataset.features_data, diabetes_dataset.labels_data, sample_weight=sample_weights
+        diabetes_dataset.features_data,
+        params={"y": diabetes_dataset.labels_data, "sample_weight": sample_weights},
```
Comment (harupy) [inline tests/evaluate/test_default_evaluator.py:191]: is this a breaking change?

Comment (WeichenXu123) [reply tests/evaluate/test_default_evaluator.py:191]: I think no, `_model_impl` is not user facing API.

File: mlflow/sklearn/__init__.py (hunk 2 above)
Comment (WeichenXu123) [reply mlflow/sklearn/__init__.py:530]: do we need this ? they all have the same pattern.

**Comment by WeichenXu123** [comment]
> I'm curious why `params` are passed in the pipeline `predict_proba` method.

It is passed here: https://github.com/mlflow/mlflow/blob/b96b36ba5be14f6b70955f72527b38aaea8cac71/mlflow/pyfunc/__init__.py#L769 

this is a design of `PyfuncModel` that allows passing additional params for `predict`

File: tests/evaluate/test_default_evaluator.py (hunk 3 above)
Comment (harupy) [reply tests/evaluate/test_default_evaluator.py:191]: but it's marked as `developer_stable`

File: tests/sklearn/test_sklearn_model_export.py (hunk 1 above)
Comment (WeichenXu123) [reply tests/sklearn/test_sklearn_model_export.py:897]: We can . but the model in https://github.com/mlflow/mlflow/issues/12445 doesn't have additional inference params to test (i.e. in this case the passed in `params` argument is None)

Comment (harupy) [reply tests/sklearn/test_sklearn_model_export.py:897]: let's use a more realistic example.

File: mlflow/sklearn/__init__.py (hunk 2 above)
Comment (harupy) [reply mlflow/sklearn/__init__.py:530]: Never mind this comment. unrelated to this PR. I'm just not a big fan of `getattr` or `setattr`.

**Comment by harupy** [comment]
I'm also curious why `predict` works fine.

**Comment by WeichenXu123** [comment]
> I'm also curious why `predict` works fine.

oh, predict doesn't work too.

File: tests/evaluate/test_default_evaluator.py (hunk 3 above)
Comment (WeichenXu123) [reply tests/evaluate/test_default_evaluator.py:191]: Yes. Does `developer_stable` this matters ? this is for developers. @BenWilson2 ?

**Comment by harupy** [comment]
> oh, predict doesn't work too.

Example?

File: tests/evaluate/test_default_evaluator.py (hunk 3 above)
Comment (harupy) [reply tests/evaluate/test_default_evaluator.py:191]: It does. It would affect developers, right?

**Comment by harupy** [comment]
Does this issue only affect sklearn flavor?

File: tests/evaluate/test_default_evaluator.py (hunk 3 above)
Comment (WeichenXu123) [reply tests/evaluate/test_default_evaluator.py:191]: Then we need to reconsider how to fix the issue. Current approach inevitably has to break this.

**Comment by WeichenXu123** [comment]
> Does this issue only affect sklearn flavor?

Yes.

**Comment by WeichenXu123** [comment]
> #12445

Oh sorry, sklearn pipeline predict method doesn't trigger this bug:

because pipeline predict function is defined like:
```
    def predict(self, X, **predict_params):
```
it has a different param name of `predict_params`

**Comment by WeichenXu123** [comment]
> > I'm also curious why `predict` works fine.
> 

yes predic works fine, because `predict` case goes anther route:
https://github.com/mlflow/mlflow/blob/b96b36ba5be14f6b70955f72527b38aaea8cac71/mlflow/sklearn/__init__.py#L524

it has a `params` arg here, so it prevents the error.

**Comment by harupy** [comment]
> it has a params arg here, so it prevents the error.

Other functions should also follow the same route?

File: mlflow/sklearn/__init__.py [hunk 4]
```diff
@@ -531,7 +542,8 @@ def predict(
         Returns:
             Model predictions.
         """
-        return self.sklearn_model.predict(data)
+        params = params or {}
+        return self.sklearn_model.predict(data, **params)
```
Comment (harupy) [inline mlflow/sklearn/__init__.py:546]: why do we need this change?

**Comment by WeichenXu123** [comment]
> > it has a params arg here, so it prevents the error.
> 
> Other prediction functions should also follow the same route?

No. Other functions like `sklearn_model.predict_proba` are directly set as `_model_impl` attribute, i.e. the master branch code does:
```
_model_impl.predict_proba = sklearn_model.predict_proba
```

**Comment by harupy** [comment]
@WeichenXu123 what's `_model_impl`? raw sklearn model? or the sklearn wrapper?

**Comment by WeichenXu123** [comment]
> @WeichenXu123 what's `_model_impl`? raw sklearn model? or the sklearn wrapper?

sklearn wrapper.

**Comment by harupy** [comment]
What would happen if we defined `predict_proba` like this? Would that suppress the issue?

```
    def predict_proba(
        self,
        data,
        params: Optional[Dict[str, Any]] = None,
    ):
        """
        Args:
            data: Model input data.
            params: Additional parameters to pass to the model for inference.

        Returns:
            Model predictions.
        """
        return self.sklearn_model.predict_proba(data)
```

**Comment by WeichenXu123** [comment]
> What would happen if we defined `predict_proba` like this? Would that suppress the issue?
> 
> ```
>     def predict_proba(
>         self,
>         data,
>         params: Optional[Dict[str, Any]] = None,
>     ):
>         """
>         Args:
>             data: Model input data.
>             params: Additional parameters to pass to the model for inference.
> 
>         Returns:
>             Model predictions.
>         """
>         return self.sklearn_model.predict_proba(data)
> ```

Yes. This can also address the user case. 

but it can't support https://github.com/mlflow/mlflow/pull/12489#discussion_r1662073405 (if this case is not important, we can ignore it for now)

**Comment by harupy** [comment]
> but it can't support https://github.com/mlflow/mlflow/pull/12489#discussion_r1662073405 (if this case is not important, we can ignore it for now)

you can address this in a follow-up.

**Comment by WeichenXu123** [comment]
Filed another PR using alternative approach to fix this: https://github.com/mlflow/mlflow/pull/12554

**Pull request closed without merging**